### ISSUE TITLE
`Readeable` streams processing API doesn't preserve order of incoming chunks

### ISSUE THREAD

**Issue by paluh** [purescript-node/purescript-node-streams#27]
What I mean by "not preserving order" can be described more easily by a real world example. [Here](https://github.com/cprussin/purescript-httpure/blob/1adbcecaaa01cab6b690abf5a0f36103ca2c1742/src/HTTPure/Body.purs#L84) we have a function from previous version of `purescript-httpure`:

```purescript
read :: HTTP.Request -> Aff.Aff String
read request = Aff.makeAff \done -> do
  let stream = HTTP.requestAsStream request
  buf <- Ref.new ""
  Stream.onDataString stream Encoding.UTF8 \str ->
    void $ Ref.modify ((<>) str) buf
  Stream.onEnd stream $ Ref.read buf >>= Either.Right >>> done
  pure Aff.nonCanceler
```
We have found that this function in case of a large request can produce body string which is concatenation of chunks but in wrong order. It seems that Javascript scheduler can interleave chunk processing handlers (because they contain multiple internal function calls) and we can't be sure that they end up in correct order.

I don't have clear proposition how to improve current API of `purescript-node-streams`, but for sure we can add description of this situation to the docs.
Maybe it would be good to extend the API and add an option to functions like `onData` or `onDataString` which would enforce sequential processing of chunks through internal use of `pause` and `resume` methods.

Of course I can try to provide a PR if we agree on what and how should be done.

What do you think?

P.S. Thanks a lot to all contributors for building these low level bindings to nodejs!

**Comment by cprussin** [comment]
Hey @garyb and @paf31 I think you two are likely the most experienced with this kind of thing, do you have any thoughts here?  @paluh and I spoke about this over in httpure issues, he submitted a fix that looks like this that _might_ make sense to go in this binding, but it's not so low level:

```purescript
foreign import aggregateChunks :: forall w. Stream.Readable w ->
                                  (Array Buffer.Buffer -> Effect.Effect Unit) ->
                                  Effect.Effect Unit
```

```js
exports.aggregateChunks = function (stream) {
  return function(onEnd) {
    return function() {
      var chunks = [];
      stream.on("data", function (chunk) {
        chunks.push(chunk);
      });
      stream.on("end", function() {
        onEnd(chunks)();
      });
    };
  }
};
```

Certainly, this resolves our issue.  But if an API like that is necessary to preserve order properly, isn't `Node.Stream.onData` then unreliable?  To be honest, I'm not really sure how this bug is manifesting--Node shouldn't be interrupting synchronous call stacks and there's nothing async in the signal handling that I'm aware of.

So, in conclusion, this is a bit out of my expertise to guide a solution, but at the least, I don't feel that the fix is anything specific to HTTPure, and whatever _is_ the right long-term solution IMO should be here.  Please advise.
